sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -222,24 +223,68 @@ object Hash {
     }
   }
 
-  // Transaction Builder class with builder methods to recursively encode nodes
-  private sealed class TransactionBuilderV1(
+  object TransactionMetadataBuilderV1 {
+    final case class Metadata(
+        actAs: ImmArray[Ref.Party],
+        commandId: Ref.CommandId,
+        submissionId: Ref.SubmissionId,
+        transactionUUID: UUID,
+        mediatorGroup: Int,
+        domainId: String,
+        ledgerEffectiveTime: Option[Time.Timestamp],
+        submissionTime: Time.Timestamp,
+        submissionSeed: Hash,
+        disclosedContracts: ImmArray[Node.Create],
+    )
+  }
+
+  /** Hashes Transaction Metadata using the V1 Hashing Scheme
+    */
+  @throws[NodeHashingError]
+  @throws[HashingError]
+  def hashTransactionMetadataV1(
+      metadata: TransactionMetadataBuilderV1.Metadata,
+      hashTracer: HashTracer = HashTracer.NoOp,
+  ): Hash = {
+    new NodeBuilderV1(hashTracer)
+      .addHashVersion(NodeHashVersion.V1)
+      .withContext("Act As Parties")(_.iterateOver(metadata.actAs)(_ addString _))
+      .withContext("Command Id")(_.addString(metadata.commandId))
+      .withContext("Submission Id")(_.addString(metadata.submissionId))
+      .withContext("Transaction UUID")(_.addString(metadata.transactionUUID.toString))
+      .withContext("Mediator Group")(_.addInt(metadata.mediatorGroup))
+      .withContext("Domain Id")(_.addString(metadata.domainId))
+      .withContext("Ledger Effective Time")(
+        _.addOptional(metadata.ledgerEffectiveTime.map(_.micros), _.addLong)
+      )
+      .withContext("Submission Time")(_.addLong(metadata.submissionTime.micros))
+      .withContext("Submission Seed")(_.addHash(metadata.submissionSeed, "Submission Seed Hash"))
+      .withContext("Disclosed Contracts")(
+        _.iterateOver(metadata.disclosedContracts)((builder, node) =>
```

Comment:
You probably want to sort those disclosed contracts before hashing them, e.g. by the contract ID. Do you also want to eliminate duplicates?

Reply:
OK let's sorted them by contract ID.

Reply:
I'll switch the type to `SortedSet`